anyio_serial/_impl.py

```diff
@@ -2,84 +2,159 @@
-        return await self._ctx.__aexit__(*tb)
-
-
-    @asynccontextmanager
-    async def _gen_ctx(self):
-        await anyio.run_sync_in_worker_thread(self._open)
-        try:
-            yield self
-        finally:
-            await self.aclose()
-
-    async def aclose(self):
-        port, self._port = self._port, None
-        if port is None:
-            return
-        await anyio.run_sync_in_worker_thread(self._close, port)
-
-
-    def _open(self):
-        self._port = serial.Serial(*self._a, **self._kw)
-
-    def _close(self, port):
-        port.close()
+from __future__ import annotations
+
+from functools import partial
+from typing import Any, Type, TypeVar
+
+import anyio
+import serial as pyserial
+from anyio import run_async_from_thread, run_sync_in_worker_thread
+from anyio.abc import ByteStream
+from anyio.streams.buffered import BufferedByteReceiveStream
+from serial.serialutil import PortNotOpenError
+
+
+T = TypeVar("T", bound="Serial")
+
+
+class Serial(ByteStream):
+    """
+    Async ByteStream for pyserial's `Serial` interface.
+
+    This object can be asynchronously iterated upon::
+
+        async with Serial(...) as serial:
+            async for data in serial:
+                print(data)
+                await serial.send(b"I got data from you!")
+
+    Args:
+        serial: a pyserial `Serial` instance.
+        max_read_size: The maximum size of a read from the serial port in the background thread.
+    """
+
+    def __init__(self, serial: pyserial.Serial, max_read_size: int = 1024) -> None:
+        self._serial = serial
+        self.max_read_size = max_read_size
+
+        self._task_group = anyio.create_task_group()
+
+        self._read_producer, self._read_consumer = anyio.create_memory_object_stream(
+            0, item_type=bytes
+        )
+        self._write_producer, self._write_consumer = anyio.create_memory_object_stream(
+            0, item_type=bytes
+        )
+        self._read_stream = BufferedByteReceiveStream(self._read_consumer)
+
+    @classmethod
+    def from_url(cls: Type[T], url: str, **kwargs: Any) -> T:
+        """
+        Open an `Serial` port with a URL.
+        See https://pyserial.readthedocs.io/en/latest/url_handlers.html
+        """
+        kwargs["do_not_open"] = True
+        serial = pyserial.serial_for_url(url, **kwargs)
+        return cls(serial)
+
+    async def __aenter__(self) -> Serial:
+        if not self._serial.is_open:
+            await run_sync_in_worker_thread(self._serial.open)
+        await self._task_group.__aenter__()
+        await self._task_group.spawn(
+            partial(run_sync_in_worker_thread, self._read_worker, cancellable=True)
+        )
+        await self._task_group.spawn(
+            partial(run_sync_in_worker_thread, self._write_worker, cancellable=True)
+        )
+        return self
+
+    async def __aexit__(self, *args: Any) -> None:
+        await self.aclose()
+        await self._task_group.__aexit__(*args)
```

Comment:
You should (a) wrap a shielded cancel scope around the `aclose` call, (b) return the value from `__aexit__`.

Reply:
`aclose()` does have a shielded cancel scope already, but I can move it to `__aexit__()`.

Reply:
Huh. I seem to have completely missed that. No matter then.